src/lib.rs

```diff
@@ -499,153 +485,217 @@ where
+            headers: None,
             data,
-            format_version: FORMAT_VERSION_V1,
             data_type,
             data_schema_version,
-            topic,
-        };
-        hedwig.validator.validate(&message)?;
-        Ok(message)
+        }
     }
 
-    /// Add custom headers to the message. This may be used to track `request_id`, for example.
-    pub fn with_headers(&mut self, headers: Headers) -> &mut Self {
-        self.metadata.headers = headers;
+    /// Overwrite the header map associated with the message.
+    ///
+    /// This may be used to track the `request_id`, for example.
+    pub fn headers(mut self, headers: Headers) -> Self {
+        self.headers = Some(headers);
         self
     }
 
-    /// Add custom id to the message. If not provided, a new uuid v4 is used.
-    pub fn with_id(&mut self, id: Uuid) -> &mut Self {
-        self.id = id;
+    /// Add a custom header to the message.
+    ///
+    /// This may be used to track the `request_id`, for example.
+    pub fn header<H, V>(mut self, header: H, value: V) -> Self
+    where
+        H: Into<String>,
+        V: Into<String>,
+    {
+        if let Some(ref mut hdrs) = self.headers {
+            hdrs.insert(header.into(), value.into());
+        } else {
+            let mut map = HashMap::new();
+            map.insert(header.into(), value.into());
+            self.headers = Some(map);
+        }
         self
     }
 
-    #[cfg(any(feature = "google", feature = "mock"))]
-    fn headers(&self) -> Headers {
-        self.metadata.headers.clone()
+    /// Add custom id to the message. If not provided, a random UUID is generated.
+    pub fn id(mut self, id: Uuid) -> Self {
+        self.id = Some(id);
+        self
+    }
+
+    fn into_schema(
+        self,
+        publisher_name: String,
+        schema: String,
+        format_version: Version,
+    ) -> Result<MessageSchema, serde_json::Error>
+    where
+        D: Serialize,
+    {
+        Ok(MessageSchema {
+            id: self.id.unwrap_or_else(Uuid::new_v4),
+            metadata: MetadataSchema {
+                timestamp: self.timestamp.as_millis(),
+                publisher: publisher_name,
+                headers: self.headers.unwrap_or_else(HashMap::new),
+            },
+            schema,
+            format_version,
+            data: serde_json::to_value(self.data)?,
+        })
     }
 }
 
+/// Additional metadata associated with a message
+#[derive(Clone, Debug, PartialEq, Serialize)]
+pub struct MetadataSchema {
+    /// The timestamp when message was created in the publishing service
+    pub timestamp: u128,
+
+    /// Name of the publishing service
+    pub publisher: String,
+
+    /// Custom headers. This may be used to track request_id, for example.
+    pub headers: Headers,
+}
+
+/// A validated message.
+///
+/// This data type is the schema or the json messages being sent over the wire.
+#[derive(Debug, Serialize)]
+pub struct MessageSchema {
```

Comment:
This doesn't need to be different from `Message` does it?

Reply:
The reason why they are separate is to separate the concerns these types handle. The `Message` is a builder whereas `MessageSchema` is the thing being serialized. Such separation aids maintainability of this code as `MessageSchema` should never change unless hedwig protocol changes, whereas it is perfectly fine for `Message` to change if we wanted to e.g. do something special about some extra data.

This also allows for validation of the message contents to not be spread out all over the place – previously the message contents were validated piecewise across the whole publish flow and the types did not reflect the state of validation at all.

Now the `Publisher::publish` takes the `MessageSchema` and so it knows that the argument it is receiving has been validated (and the data has been encoded to JSON) already.

Reply:
Well, I suppose my point was that the class is simple enough that you don't need to separate out the transport from data packet. I agree that the validation should be more "centralized". For example, this class could be converted into an enum with state machine that transitions from "creating a message" to "validated message". After validation no changes can be made to the data.

As it's written currently, the class name isn't very descriptive. Can we call this `ValidatedMessage` or `MessageTransport` .. something along those lines?

Reply:
> For example, this class could be converted into an enum with state machine that transitions from "creating a message" to "validated message". After validation no changes can be made to the data.

I strongly disagree with that sentiment. Having multiple types, however similar they are, to represent different type states (and thus to enforce correct usage of them at the compile time) is a well known and widely used pattern. Using anything less usually means the invariants can only be enforced at run time.

---

The name here comes from the fact that this type represents the hedwig on-wire data 1:1. Thus "schema". I don’t have a strong opinion, though, so whatever, I’ll name it a validated message instead.

Reply:
FWIW we could do with just a single type while still retaining the compile-time guarantees, but in that case this library will want some more rework of its overall structure (esp. pertaining to the `Validator` and `Publisher`).

Reply:
I mean its still different types ([this](https://hoverbear.org/2016/10/12/rust-state-machine-pattern/) is the pattern I'm referring to), but this is good enough.